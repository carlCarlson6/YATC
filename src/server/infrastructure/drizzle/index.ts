import * as baseSchema from "src/server/infrastructure/drizzle/base.drizzle.schema";
import * as followsSchema from "src/server/user/follows/follows.drizzle.schema"
import * as emojisSchema from "src/server/publish-emojeet/emojis.drizzle.schema";
import { sql } from '@vercel/postgres';
//import { drizzle } from 'drizzle-orm/vercel-postgres';
import { drizzle } from 'drizzle-orm/postgres-js';

import postgres from 'postgres';
import { env } from "src/env.mjs";


export const drizzleSchema = {
	...baseSchema,
  ...followsSchema,
  ...emojisSchema,
};

//export const drizzleDb = drizzle(sql, { schema: drizzleSchema } );

export const drizzleDb = drizzle(postgres(env.NEON_DB), { schema: drizzleSchema } );
export type DrizzleDb = typeof drizzleDb;