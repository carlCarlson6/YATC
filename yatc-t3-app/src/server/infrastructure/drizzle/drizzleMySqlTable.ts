import { mysqlTableCreator } from "drizzle-orm/mysql-core";
import { env } from "yact/env.mjs";

export const drizzleMySqlTable = mysqlTableCreator((name) => `${env.APP_ENV}yatct3app${name}`);