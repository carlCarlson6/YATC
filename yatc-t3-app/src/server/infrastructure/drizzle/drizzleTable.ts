import { pgTableCreator } from 'drizzle-orm/pg-core'

export const drizzleTable = pgTableCreator((name) => `yatct3app${name}`);