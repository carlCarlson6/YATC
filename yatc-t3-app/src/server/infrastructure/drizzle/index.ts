import { Client } from "@planetscale/database";
import { drizzle } from "drizzle-orm/planetscale-serverless";
import { env } from "yact/env.mjs";
import * as baseSchema from "yact/server/infrastructure/drizzle/base.drizzle.schema";
import * as tweetSchema from "yact/server/send-tweet/tweet.drizzle.schema";
import * as followsSchema from "yact/server/user/follow/follow.drizzle.schema"

const schema = {
	...baseSchema,
  ...tweetSchema,
  ...followsSchema,
};

export const drizzleDb = drizzle(
  new Client({
    url: env.DATABASE_URL,
  }).connection(),
  { schema }
);

export type DrizzleDb = typeof drizzleDb;