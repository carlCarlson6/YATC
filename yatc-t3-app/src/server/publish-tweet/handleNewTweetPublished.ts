import { eq } from "drizzle-orm";
import { type DrizzleDb } from "../infrastructure/drizzle";
import { usersTable } from "../infrastructure/drizzle/base.drizzle.schema";
import { followsTable } from "../user/follows/follow.drizzle.schema";
import { type SendUpdateUserTimeline } from "../timeline/sendUpdateUserTimeline";
import type { updateUserTimeline } from "../timeline/handleUpdateUserTimeline";

export const handleNewTweetPublished = ({loadFollower, publish, updateTimeline}: {
  loadFollower: (userId: string) => Promise<string[]>, 
  publish: SendUpdateUserTimeline,
  updateTimeline: ReturnType<typeof updateUserTimeline>
}) => async (event: {
  tweetId: string, 
  publishedBy: string
}) => {
  await updateTimeline({userId: event.publishedBy, tweetId: event.tweetId});
  const followers = await loadFollower(event.publishedBy);
  const publishAction = followers.map(follower => publish({tweetId: event.tweetId, userId: follower}));
  await Promise.all(publishAction);
}

export const loadFollowersFromDrizzleDb = (db: DrizzleDb) => async (userId: string) => {
  const result = await db
    .select({
      follower: usersTable.id
    })
    .from(followsTable)
    .where(eq(followsTable.isFollowingUserId, userId))
    .rightJoin(usersTable, eq(followsTable.isFollowingUserId, usersTable.id))
    .execute();
  return result.map(({follower}) => follower);
}

