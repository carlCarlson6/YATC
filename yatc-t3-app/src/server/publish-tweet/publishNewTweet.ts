import type { DrizzleDb } from "../infrastructure/drizzle";
import type { QStashPublisher } from "../infrastructure/qstash";
import type { User } from "../user/user";
import { tweetsTable, type TweetEntity } from "./tweet.drizzle.schema";

export const publishNewTweet = ({storeTweet, sendNewTweetPublished}: {
  storeTweet: (tweet: TweetEntity) => Promise<void>,
  sendNewTweetPublished: (tweet: TweetEntity) => Promise<void>,
}) => async (
  userPublisher: User,
  tweetText: string
) => {
  const tweet = {
    id: crypto.randomUUID(),
    text: tweetText,
    publishedBy: userPublisher.id,
    publishedAt: Date.now()/1000,
  };

  await storeTweet(tweet);
  await sendNewTweetPublished(tweet);

  return ({
    ...tweet,
    user: userPublisher
  });
};

export const storeTweetWithDrizzle = (db: DrizzleDb) => async (tweet: TweetEntity) => {
  await db.insert(tweetsTable).values(tweet).execute();
}

export const sendNewTweetPublishedWithQStash = (qstash: QStashPublisher, appUrl: string) => async (tweet: TweetEntity) => {
  console.log("publishing event", "tweet-published", "to", `${appUrl}/api/qstash`);
  const result = await qstash(`${appUrl}/api/qstash`, {
    type: "tweet-published",
    payload: JSON.stringify({ tweetId: tweet.id, publishedBy: tweet.publishedBy })
  });
  console.log("result of send message", result);
}