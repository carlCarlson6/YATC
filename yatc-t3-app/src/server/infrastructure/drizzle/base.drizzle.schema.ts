import { relations, sql } from "drizzle-orm";
import {
  index,
  integer,
  primaryKey,
  text,
  timestamp,
  varchar,
} from "drizzle-orm/pg-core";
import { tweetsTable } from "src/server/send-tweet/tweet.drizzle.schema";
import { drizzleTable } from "./drizzleTable";
import type { AdapterAccount } from "node_modules/next-auth/adapters";

export const usersTable = drizzleTable("user", {
  id: varchar("id", { length: 255 }).notNull().primaryKey(),
  name: varchar("name", { length: 255 }),
  email: varchar("email", { length: 255 }).notNull(),
  emailVerified: timestamp("emailVerified", { mode: "date", }).default(sql`CURRENT_TIMESTAMP(3)`),
  image: varchar("image", { length: 255 }),
});

export const usersTableRelations = relations(usersTable, ({ many }) => ({
  accounts: many(accountsTable),
  tweets: many(tweetsTable)
}));

export const accountsTable = drizzleTable(
  "account",
  {
    userId: varchar("userId", { length: 255 }).notNull(),
    type: varchar("type", { length: 255 })
      .$type<AdapterAccount["type"]>()
      .notNull(),
    provider: varchar("provider", { length: 255 }).notNull(),
    providerAccountId: varchar("providerAccountId", { length: 255 }).notNull(),
    refresh_token: text("refresh_token"),
    access_token: text("access_token"),
    expires_at: integer("expires_at"),
    token_type: varchar("token_type", { length: 255 }),
    scope: varchar("scope", { length: 255 }),
    id_token: text("id_token"),
    session_state: varchar("session_state", { length: 255 }),
  },
  (account) => ({
    compoundKey: primaryKey(account.provider, account.providerAccountId),
    userIdIdx: index("userId_idx").on(account.userId),
  })
);

export const accountsTableRelations = relations(accountsTable, ({ one }) => ({
  user: one(usersTable, { fields: [accountsTable.userId], references: [usersTable.id] }),
}));

export const sessionsTable = drizzleTable(
  "session",
  {
    sessionToken: varchar("sessionToken", { length: 255 })
      .notNull()
      .primaryKey(),
    userId: varchar("userId", { length: 255 }).notNull(),
    expires: timestamp("expires", { mode: "date" }).notNull(),
  },
  (session) => ({
    userIdIdx: index("userId_idx").on(session.userId),
  })
);

export const sessionsTableRelations = relations(sessionsTable, ({ one }) => ({
  user: one(usersTable, { fields: [sessionsTable.userId], references: [usersTable.id] }),
}));

export const verificationTokensTable = drizzleTable(
  "verificationtoken",
  {
    identifier: varchar("identifier", { length: 255 }).notNull(),
    token: varchar("token", { length: 255 }).notNull(),
    expires: timestamp("expires", { mode: "date" }).notNull(),
  },
  (vt) => ({
    compoundKey: primaryKey(vt.identifier, vt.token),
  })
);
