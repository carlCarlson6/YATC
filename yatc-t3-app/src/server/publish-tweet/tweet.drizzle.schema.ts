import { relations } from "drizzle-orm";
import { usersTable } from "../infrastructure/drizzle/base.drizzle.schema"
import { decimal, varchar } from "drizzle-orm/pg-core";
import { drizzleTable } from "../infrastructure/drizzle/drizzleTable";

export type TweetEntity = typeof tweetsTable.$inferSelect;

export const tweetsTable = drizzleTable("tweet", {
  id:           varchar("id", { length: 255 }).notNull().primaryKey(),
  text:         varchar("text", { length: 280 }).notNull(),
  publishedBy:  varchar("publishedBy", { length: 255 }).notNull(),
  publishedAt:  decimal("publishedAt").notNull(),
});

export const tweetsTableRelations = relations(tweetsTable, ({one}) => ({
  user: one(usersTable, { fields: [tweetsTable.publishedBy], references: [usersTable.id] })
}));