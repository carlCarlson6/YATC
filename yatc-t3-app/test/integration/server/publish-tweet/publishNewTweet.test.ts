import { expect,vi, test } from 'vitest';
import { publishNewTweet, storeTweetWithDrizzle } from 'src/server/publish-tweet/publishNewTweet';
import { TweetEntity } from 'src/server/publish-tweet/tweet.drizzle.schema';
import { setupDockerTestDb } from 'test/helpers/setupTestingDb';

test("GivenTweet_WhenPublishTweet_ThenANewTweetIsReturned", async () => {
  const {db} = await setupDockerTestDb();
  
  const newTweet = await publishNewTweet({
    storeTweet: storeTweetWithDrizzle(db),
    sendNewTweetPublished: vi.fn((_: TweetEntity) => Promise.resolve())
  })({ id: "some-user-id", name: "some-user", avatar: "some-image.png" }, "some-text");

  expect(newTweet).toMatchSnapshot({id: expect.any(String)})
});

test("GivenTweet_WhenPublishTweet_ThenANewTweetIsStored", async () => {
  const {db} = await setupDockerTestDb();

  await publishNewTweet({
    storeTweet: storeTweetWithDrizzle(db),
    sendNewTweetPublished: vi.fn((_: TweetEntity) => Promise.resolve())
  })({ id: "some-user-id", name: "some-user", avatar: "some-image.png" }, "some-text");
  
  const storedTweet = await db.query.tweetsTable.findFirst().execute();
  expect(storedTweet).toMatchSnapshot({id: expect.any(String)})
});

test("GivenTweet_WhenPublishTweet_ThenSendNewTweetPublishedIsCalled", async () => {
  const {db} = await setupDockerTestDb();
  const sendNewTweetPublished =  vi.fn((tweet: TweetEntity) => Promise.resolve());

  await publishNewTweet({
    storeTweet: storeTweetWithDrizzle(db),
    sendNewTweetPublished
  })({ id: "some-user-id", name: "some-user", avatar: "some-image.png" }, "some-text");

  expect(sendNewTweetPublished).toHaveBeenCalledTimes(1);
});