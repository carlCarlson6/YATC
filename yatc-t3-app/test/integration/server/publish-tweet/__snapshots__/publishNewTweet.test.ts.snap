// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GivenTweet_WhenPublishTweet_ThenANewTweetIsReturned 1`] = `
{
  "id": Any<String>,
  "publishedAt": 1701561649.243,
  "publishedBy": "some-user-id",
  "text": "some-text",
  "user": {
    "avatar": "some-image.png",
    "id": "some-user-id",
    "name": "some-user",
  },
}
`;

exports[`GivenTweet_WhenPublishTweet_ThenANewTweetIsStored 1`] = `
{
  "id": Any<String>,
  "publishedAt": "1701561650.365",
  "publishedBy": "some-user-id",
  "text": "some-text",
}
`;
